### PR TITLE
tailcfg: add DNSConfig.SetDNSDomains

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #1235

Signed-off-by: Brad Fitzpatrick <bradfitz@tailscale.com>

### CODE REVIEW

**Comment by bradfitz** [comment]
Thoughts?

**Comment by bradfitz** [comment]
Don't want it top-level in MapResponse as I don't want to add new minnification code to control for it. DNSConfig is already dup-suppressed.

File: tailcfg/tailcfg.go [hunk 1]
```diff
@@ -873,6 +874,10 @@ type DNSConfig struct {
 
 	// PerDomain is not set by the control server, and does nothing.
 	PerDomain bool `json:",omitempty"`
+
+	// SetDNSDomains are the set of DNS names that are permitted to
```
Comment (danderson) [inline tailcfg/tailcfg.go:878]: I'm not sure about the field name. SetDNSDomains sounds like an imperative command for the client to set some DNS domains, not "these are the domains you can choose to request stuff on". We also don't guarantee that you can set whatever records you want on these names, it's just a hint that "if you're trying to do LE, these are the ones that are likely to succeed".

I don't have better ideas off the top of my head, the DNS record-setting stuff in tailcfg is so generic that it's hard to come up with reasonable names :/

**Review by danderson — APPROVED** [review]

File: tailcfg/tailcfg.go (hunk 1 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go:878]: ValidDNSCertDomain ?

Comment (bradfitz) [reply tailcfg/tailcfg.go:878]: ValidCertDomains?

Comment (normanr) [reply tailcfg/tailcfg.go:878]: What's the motivation for this field? Does it get set by the tailscale administrator to limit which machines may set TXT records, or does it limit which "sub-names" are allowed (eg: `_acme-challenge`), or is it both? Should it allow wildcards, eg: `_acme-challenge.*` (to allow any machine to use LE), or `*.some.awesome.machine` (which can set a TXT record for _any_ sub-name)

Comment (bradfitz) [reply tailcfg/tailcfg.go:878]: @normanr, we haven't decided what subdomain structure we're going to use for LetsEncrypt certs. There are various concerns about customer hostnames/emails leaking to CT logs if we do it with our current MagicDNS names. So we might make this opt-in with separate hostnames. This is giving us a place to advertise that policy from the server to the client later, so we can update it server side and not require client updates.

Comment (bradfitz) [reply tailcfg/tailcfg.go:878]: Decision from @danderson: just `CertDomains`.

**Pull request merged**